### PR TITLE
Prevent deadlock between observe and replace methods in ADS Observers

### LINKED ISSUE #30523: Workbench can hang if close plot during ADS update
### Expected behavior
You should be able to open and close plots while the ADS is updating

### Actual behavior
If you close a plot while the ADS is refreshing the workspace, it can hang workbench. I think there also may be errors in the log opening a plot on a workspace that is being updated.

### Steps to reproduce the behavior
- In the Settings, set the default instrument to TEST_LIVE
- Run the algorithm FakeISISHistoDAE
- Go to Load->Live data and start live data monitoring with instrument `ISIS_Histogram`, an update interval of `0.1`, accumulation method `Replace` and no processing or post-processing.
- Open 10 plots on the live data workspace.
- Close all the plots in quick succession.

### Platforms affected
The hang only seems to occur on Windows. I have seen errors in the log in Ubuntu. It's an existing problem in v5.1.1

### PR DESCRIPTION
**Pull request by joseph-torsney**
Fixes a deadlock that occurred when one of the observe methods of the ADS observer are called from python while a handle was being executed. This is fixed by releasing the GIL in the boost python layer before we call an observe method.

**To test:**
Two ways. Either the method described in [#30523](https://github.com/mantidproject/mantid/issues/30523) or:
1. Add a `time.sleep(2)` line to [`FigureManagerADSObserver.replaceHandle`](). This will cause the handle to sleep for two seconds.
2. Run the following script
```
from mantid.simpleapi import *

ws = CreateSampleWorkspace()
while True:
    ws = Scale(ws, 1.001, "Multiply")
```
3. Plot a spectrum from the workspace
4. Close the plot while the status bar indicates `Multiplying factor... 50%`
5. Workbench should not hang.

Fixes #30523

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserverAdapter.cpp [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #include "MantidPythonInterface/api/AnalysisDataServiceObserverAdapter.h"
 #include "MantidAPI/AnalysisDataServiceObserver.h"
 #include "MantidPythonInterface/core/CallMethod.h"
+#include "MantidPythonInterface/core/ReleaseGlobalInterpreterLock.h"
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserverAdapter.cpp:10]: I think we have an extra include here that's no longer necessary?

File: qt/applications/workbench/workbench/plotting/figuremanager.py [hunk 2]
```diff
@@ -333,6 +335,7 @@ def destroy(self, *args):
             # It seems that when the python session is killed,
             # Gcf can get destroyed before the Gcf.destroy
             # line is run, leading to a useless AttributeError.
+        self.deleteLater()
```
Comment (martyngigg) [inline qt/applications/workbench/workbench/plotting/figuremanager.py:338]: Could I ask why this was necessary? Was the figuremanager not being deleted?

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: qt/applications/workbench/workbench/plotting/figuremanager.py (hunk 2 above)
Comment (joseph-torsney) [reply qt/applications/workbench/workbench/plotting/figuremanager.py]: Thanks for spotting these, I think this is left over from when I was trying a few different things. In any case it's not needed as this is in the destroy method which is called when the object is about to be deleted anyway!

File: Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp [hunk 3]
```diff
@@ -15,26 +16,71 @@ using namespace Mantid::API;
 using namespace Mantid::PythonInterface;
 using namespace boost::python;
 
+namespace {
+void observeAll(AnalysisDataServiceObserver &self, bool turnOn) {
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp:20]: I was wondering whether we might be able to reduce the code duplication here using a more generic function and `std::bind`, for example:

```cpp
#include <functional>
#include <iostream>

// Example class as a substitute for AnalysisDataServiceObserver
struct Observer {

  void hello(bool on) {
     std::cerr << "Hello\n";
  }

  void bye(bool on) {
     std::cerr << "Bye\n";
  }

};

// Function pointer to a method with this signature
using ObserverMethod = void(Observer::*)(bool);

// Single function that calls the given method
void callReleasingGIL(Observer &self, bool on, ObserverMethod method) {
  // RELEASE GIL
  (self.*method)(on);
}

int main() {
  // use std::bind to create functions that are bound to the correct member function
  using namespace std::placeholders;
  std::bind(callReleasingGIL, _1, _2, &Observer::hello);
  std::bind(callReleasingGIL, _1, _2, &Observer::bye);

  // I am imagining something like the following below:
  // .def("observeAll", std::bind(callReleasingGIL, _1, _2, &AnalysisDataServiceObserver::observeAll);

}
```

An example on  [compiler explorer](https://gcc.godbolt.org/z/o3PzKGq5b)

Comment (joseph-torsney) [reply Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp:20]: Ah nice thank you! I'll have a go

File: Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp [hunk 4]
```diff
@@ -7,34 +7,84 @@
...
+} // namespace
 
 void export_AnalysisDataServiceObserver() {
   boost::python::class_<AnalysisDataServiceObserver, bases<>, AnalysisDataServiceObserverAdapter, boost::noncopyable>(
       "AnalysisDataServiceObserver", "Observes AnalysisDataService notifications: all only")
-      .def("observeAll", &AnalysisDataServiceObserverAdapter::observeAll, (arg("self"), arg("on")),
+      .def("observeAll",
+           make_function(std::bind(callReleasingGIL, _1, _2, &AnalysisDataServiceObserver::observeAll),
+                         default_call_policies(), (arg("self"), arg("on")),
+                         boost::mpl::vector<void, AnalysisDataServiceObserver &, bool>()),
```
Comment (martyngigg) [inline Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp:47]: This is really nice. Can I ask for one more thing to help with readability? We use the same `boost::mpl::vector` everywhere - can we alias it as well so something like

```cpp
using ObserverMethodSignature = boost::mpl::vector<void, AnalysisDataServiceObserver &, bool>
```

above and then you can just do 

```cpp
default_call_policies(), (arg("self"), arg("on")), ObserverMethodSignature()
```

so it's clearer to the caller what that structure is meant for?

Comment (joseph-torsney) [reply Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp:47]: Yeah that looks much better!

**Review by martyngigg — DISMISSED** [review]
This is a nice clean fix for the deadlock issue.

File: Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp [hunk 5]
```diff
@@ -7,34 +7,76 @@
...
+           make_function(std::bind(callReleasingGIL, _1, _2, &AnalysisDataServiceObserver::observeUnGroup),
+                         default_call_policies(), (arg("self"), arg("on")), ObserverMethodSignature()),
            "Observe AnalysisDataService for a group being removed from the ADS")
-      .def("observeGroupUpdate", &AnalysisDataServiceObserverAdapter::observeGroupUpdate, (arg("self"), arg("on")),
+      .def("observeGroupUpdate",
+           make_function(std::bind(callReleasingGIL, _1, _2, &AnalysisDataServiceObserver::observeGroupUpdate),
+                         default_call_policies(), (arg("self"), arg("on")), ObserverMethodSignature()),
            "Observe AnalysisDataService for a group being updated by being "
            "added to or removed from");
 }
```
Comment (gvardany) [inline Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp:82]: Is there a missing newline here?

Comment (martyngigg) [reply Framework/PythonInterface/mantid/api/src/Exports/AnalysisDataServiceObserver.cpp:82]: Good catch. Sounds like something we should add to the pre-commit checks.

**Review by martyngigg — APPROVED** [review]

**Pull request merged**